From 55a4fd744886baaa39f73d8e08b96a5d8fe818d0 Mon Sep 17 00:00:00 2001
From: cnjn <1779895255@qq.com>
Date: Sat, 22 Aug 2026 20:00:21 +0800
Subject: [PATCH 12/16] dt-bindings: net: add ZX279133 NPPT and RTL8372N
 integration

Consolidate 3 validated ZX279133 development commits into one functional stage.

Original range: dbbc1ee6b435d6fbfaadb48abb0794a7e42e7198..39d05903fecadc37ccbbf10ff207da91923eeaf8

Signed-off-by: cnjn <1779895255@qq.com>
---
 .../net/dsa/zte,zx279133-rtl8372n.yaml        |  93 ++++++++
 .../bindings/net/zte,zx279133-nppt.yaml       | 198 ++++++++++++++++--
 .../soc/zte/zte,zx279133-idm-cci.yaml         |  39 ++++
 .../soc/zte/zte,zx279133-pon-route.yaml       |  39 ++++
 MAINTAINERS                                   |   2 +
 5 files changed, 357 insertions(+), 14 deletions(-)
 create mode 100644 Documentation/devicetree/bindings/net/dsa/zte,zx279133-rtl8372n.yaml
 create mode 100644 Documentation/devicetree/bindings/soc/zte/zte,zx279133-idm-cci.yaml
 create mode 100644 Documentation/devicetree/bindings/soc/zte/zte,zx279133-pon-route.yaml

diff --git a/Documentation/devicetree/bindings/net/dsa/zte,zx279133-rtl8372n.yaml b/Documentation/devicetree/bindings/net/dsa/zte,zx279133-rtl8372n.yaml
new file mode 100644
index 0000000000000000000000000000000000000000..2c41edb435107955380be678aade1cd6e2e7bb5f
--- /dev/null
+++ b/Documentation/devicetree/bindings/net/dsa/zte,zx279133-rtl8372n.yaml
@@ -0,0 +1,93 @@
+# SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+%YAML 1.2
+---
+$id: http://devicetree.org/schemas/net/dsa/zte,zx279133-rtl8372n.yaml#
+$schema: http://devicetree.org/meta-schemas/core.yaml#
+
+title: RTL8372N switch on the ZTE ZX279133 NPPT LAN path
+
+maintainers:
+  - cnjn <1779895255@qq.com>
+
+description:
+  The RTL8372N CPU port connects to ZX279133 XMAC0. Switch management uses the
+  SoC MDIO controller's indirect SMI transport, represented by zte,mdio-handle.
+  The CPU conduit is a dedicated netdev owned by the NPPT Ethernet controller
+  rather than the WAN netdev.
+
+properties:
+  compatible:
+    const: zte,zx279133-rtl8372n
+
+  zte,mdio-handle:
+    $ref: /schemas/types.yaml#/definitions/phandle
+    description: MDIO child implementing the RTL8372N indirect SMI transport
+
+  reset-gpios:
+    maxItems: 1
+    description: Active-low RTL8372N hardware reset
+
+allOf:
+  - $ref: dsa.yaml#/$defs/ethernet-ports
+
+required:
+  - compatible
+  - zte,mdio-handle
+  - reset-gpios
+
+unevaluatedProperties: false
+
+examples:
+  - |
+    #include <dt-bindings/gpio/gpio.h>
+
+    mdio {
+        #address-cells = <1>;
+        #size-cells = <0>;
+
+        rtl8372n_smi: switch-controller@1d {
+            compatible = "realtek,rtl8372n";
+            reg = <0x1d>;
+        };
+    };
+
+    ethernet {
+        #address-cells = <1>;
+        #size-cells = <0>;
+
+        lan_cpu: ethernet-lan-cpu {
+            compatible = "zte,zx279133-lan-conduit";
+        };
+
+        switch {
+            compatible = "zte,zx279133-rtl8372n";
+            zte,mdio-handle = <&rtl8372n_smi>;
+            reset-gpios = <&gpio1 1 GPIO_ACTIVE_LOW>;
+
+            ports {
+                #address-cells = <1>;
+                #size-cells = <0>;
+
+                port@7 {
+                    reg = <7>;
+                    label = "lan1";
+                    phy-mode = "internal";
+                    fixed-link {
+                        speed = <2500>;
+                        full-duplex;
+                    };
+                };
+
+                port@8 {
+                    reg = <8>;
+                    ethernet = <&lan_cpu>;
+                    phy-mode = "10gbase-r";
+                    fixed-link {
+                        speed = <10000>;
+                        full-duplex;
+                    };
+                };
+            };
+        };
+    };
+...
diff --git a/Documentation/devicetree/bindings/net/zte,zx279133-nppt.yaml b/Documentation/devicetree/bindings/net/zte,zx279133-nppt.yaml
index ab2454ffcb551fca0f7660c5ad97ed09a013b87c..0c0602637721d84e5c2f97822025f0851fb8e229 100644
--- a/Documentation/devicetree/bindings/net/zte,zx279133-nppt.yaml
+++ b/Documentation/devicetree/bindings/net/zte,zx279133-nppt.yaml
@@ -4,54 +4,224 @@
 $id: http://devicetree.org/schemas/net/zte,zx279133-nppt.yaml#
 $schema: http://devicetree.org/meta-schemas/core.yaml#
 
-title: ZTE ZX279133 NPPT packet processor
+title: ZTE ZX279133 NPPT Ethernet controller
 
 maintainers:
   - cnjn <1779895255@qq.com>
 
 description:
-  The ZX279133 NPPT node describes the aggregate resources of the packet
-  processor block.
+  The ZX279133 NPPT Ethernet controller contains the XMAC control path and IDM
+  packet DMA engine. PCS, SerDes, and external PHY resources are represented by
+  their respective subsystem nodes rather than additional NPPT register ranges.
+
+allOf:
+  - $ref: ethernet-controller.yaml#
+  - if:
+      required:
+        - switch
+    then:
+      properties:
+        phys:
+          minItems: 2
+        phy-names:
+          minItems: 2
+      required:
+        - ethernet-lan-cpu
+        - zte,lan-pcs-handle
 
 properties:
   $nodename:
-    pattern: "^packet-processor@[0-9a-f]+$"
+    pattern: "^ethernet@[0-9a-f]+$"
 
   compatible:
     const: zte,zx279133-nppt
 
   reg:
-    maxItems: 1
+    items:
+      - description: NPPT traffic manager, MAC control, and IDM registers
+      - description: PPS search-engine and packet-processing registers
+
+  reg-names:
+    items:
+      - const: nppt
+      - const: pps
 
   interrupts:
-    maxItems: 1
+    items:
+      - description: NPPT aggregate interrupt
+      - description: IDM CPU receive interrupt
+      - description: IDM companion-path interrupt
+      - description: IDM buffer-release interrupt
+      - description: IDM local-test interrupt
+
+  interrupt-names:
+    items:
+      - const: nppt
+      - const: rx
+      - const: idm
+      - const: buffer-release
+      - const: local-test
 
   clocks:
-    maxItems: 1
+    items:
+      - description: IDM AXI clock
+      - description: Traffic-manager AXI clock
+      - description: NPPT APB clock
+      - description: SMAC working clock
+      - description: WOE1 working clock
+      - description: XMAC working clock
 
   clock-names:
-    const: wclk
+    items:
+      - const: idm-aclk
+      - const: tm-aclk
+      - const: pclk
+      - const: smac-wclk
+      - const: woe1-wclk
+      - const: mac-wclk
+
+  memory-region:
+    items:
+      - description: Reserved packet-buffer memory owned by the BMU
+      - description: Reserved DDR hash-table memory owned by the search engine
+      - description: Reserved descriptor memory owned by the IDM
+
+  memory-region-names:
+    items:
+      - const: bmu
+      - const: se-hash
+      - const: idm
+
+  pcs-handle:
+    maxItems: 1
+
+  phys:
+    minItems: 1
+    maxItems: 2
+
+  phy-names:
+    minItems: 1
+    items:
+      - const: serdes
+      - const: lan-serdes
+
+  zte,lan-pcs-handle:
+    $ref: /schemas/types.yaml#/definitions/phandle
+    description: XPCS0 handle for the optional XMAC0 LAN conduit
+
+  zte,pon-route:
+    $ref: /schemas/types.yaml#/definitions/phandle
+    description: Syscon containing the PON SerDes to XMAC1 route control
+
+  zte,idm-cci:
+    $ref: /schemas/types.yaml#/definitions/phandle
+    description: Syscon containing the two IDM CCI access-control words
+
+  ethernet-lan-cpu:
+    type: object
+    description: Dedicated DSA conduit netdev for the XMAC0 LAN path
+    properties:
+      compatible:
+        const: zte,zx279133-lan-conduit
+    required:
+      - compatible
+    additionalProperties: false
+
+  switch:
+    $ref: /schemas/net/dsa/zte,zx279133-rtl8372n.yaml
+    description: RTL8372N switch attached to the XMAC0 LAN path
 
 required:
   - compatible
   - reg
+  - reg-names
   - interrupts
+  - interrupt-names
   - clocks
   - clock-names
+  - memory-region
+  - memory-region-names
+  - pcs-handle
+  - phys
+  - phy-names
+  - zte,pon-route
+  - zte,idm-cci
 
-additionalProperties: false
+unevaluatedProperties: false
 
 examples:
   - |
     #include <dt-bindings/clock/zte,zx279133-topcrm.h>
     #include <dt-bindings/interrupt-controller/arm-gic.h>
 
-    packet-processor@19000000 {
+    xpcs1: ethernet-pcs@1b000000 {
+        compatible = "snps,dw-xpcs";
+        reg = <0x1b000000 0x800000>;
+    };
+
+    pon_route: syscon@17000080 {
+        compatible = "zte,zx279133-pon-route", "syscon";
+        reg = <0x17000080 0x4>;
+    };
+
+    idm_cci: syscon@10e00078 {
+        compatible = "zte,zx279133-idm-cci", "syscon";
+        reg = <0x10e00078 0x8>;
+    };
+
+    pon_serdes: phy {
+        #phy-cells = <0>;
+    };
+
+    reserved-memory {
+        #address-cells = <1>;
+        #size-cells = <1>;
+        ranges;
+
+        se_hash_reserved: buffer@9e100000 {
+            reg = <0x9e100000 0x1000000>;
+            no-map;
+        };
+
+        idm_reserved: buffer@9d700000 {
+            reg = <0x9d700000 0xa00000>;
+            no-map;
+        };
+
+        bmu_reserved: buffer@9f100000 {
+            reg = <0x9f100000 0xf00000>;
+            no-map;
+        };
+    };
+
+    ethernet@19000000 {
         compatible = "zte,zx279133-nppt";
-        reg = <0x19000000 0x1000000>;
-        interrupts = <GIC_SPI 5 IRQ_TYPE_LEVEL_HIGH>;
-        clocks = <&topcrm ZX279133_TOPCRM_CLK_PON_WOE1_WCLK>;
-        clock-names = "wclk";
+        reg = <0x19000000 0x1000000>,
+              <0x18000000 0x1000000>;
+        reg-names = "nppt", "pps";
+        interrupts = <GIC_SPI 5 IRQ_TYPE_LEVEL_HIGH>,
+                     <GIC_SPI 1 IRQ_TYPE_LEVEL_HIGH>,
+                     <GIC_SPI 2 IRQ_TYPE_LEVEL_HIGH>,
+                     <GIC_SPI 3 IRQ_TYPE_LEVEL_HIGH>,
+                     <GIC_SPI 4 IRQ_TYPE_LEVEL_HIGH>;
+        interrupt-names = "nppt", "rx", "idm", "buffer-release",
+                          "local-test";
+        clocks = <&topcrm ZX279133_TOPCRM_CLK_PON_IDM_ACLK>,
+                 <&topcrm ZX279133_TOPCRM_CLK_PON_TM_ACLK>,
+                 <&topcrm ZX279133_TOPCRM_CLK_PON_PCLK>,
+                 <&topcrm ZX279133_TOPCRM_CLK_PON_SMAC_WCLK>,
+                 <&topcrm ZX279133_TOPCRM_CLK_PON_WOE1_WCLK>,
+                 <&topcrm ZX279133_TOPCRM_CLK_PON_MAC_WCLK>;
+        clock-names = "idm-aclk", "tm-aclk", "pclk", "smac-wclk",
+                      "woe1-wclk", "mac-wclk";
+        memory-region = <&bmu_reserved>, <&se_hash_reserved>, <&idm_reserved>;
+        memory-region-names = "bmu", "se-hash", "idm";
+        pcs-handle = <&xpcs1>;
+        phys = <&pon_serdes>;
+        phy-names = "serdes";
+        zte,pon-route = <&pon_route>;
+        zte,idm-cci = <&idm_cci>;
+        phy-mode = "2500base-x";
         status = "disabled";
     };
 
diff --git a/Documentation/devicetree/bindings/soc/zte/zte,zx279133-idm-cci.yaml b/Documentation/devicetree/bindings/soc/zte/zte,zx279133-idm-cci.yaml
new file mode 100644
index 0000000000000000000000000000000000000000..b4c597fc9de3e5c4ef568a210aa2e4aea81b95b5
--- /dev/null
+++ b/Documentation/devicetree/bindings/soc/zte/zte,zx279133-idm-cci.yaml
@@ -0,0 +1,39 @@
+# SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+%YAML 1.2
+---
+$id: http://devicetree.org/schemas/soc/zte/zte,zx279133-idm-cci.yaml#
+$schema: http://devicetree.org/meta-schemas/core.yaml#
+
+title: ZTE ZX279133 IDM CCI access control
+
+maintainers:
+  - cnjn <1779895255@qq.com>
+
+description:
+  Two system-control words configure the IDM interfaces used by the network
+  DMA engine. The Ethernet driver snapshots them before enabling the vendor
+  Linux value and restores them after the IDM path is stopped.
+
+properties:
+  compatible:
+    items:
+      - const: zte,zx279133-idm-cci
+      - const: syscon
+
+  reg:
+    maxItems: 1
+
+required:
+  - compatible
+  - reg
+
+additionalProperties: false
+
+examples:
+  - |
+    syscon@10e00078 {
+        compatible = "zte,zx279133-idm-cci", "syscon";
+        reg = <0x10e00078 0x8>;
+    };
+
+...
diff --git a/Documentation/devicetree/bindings/soc/zte/zte,zx279133-pon-route.yaml b/Documentation/devicetree/bindings/soc/zte/zte,zx279133-pon-route.yaml
new file mode 100644
index 0000000000000000000000000000000000000000..88eb503f6d4ce9a88146b847fe2aa7eb565eebba
--- /dev/null
+++ b/Documentation/devicetree/bindings/soc/zte/zte,zx279133-pon-route.yaml
@@ -0,0 +1,39 @@
+# SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+%YAML 1.2
+---
+$id: http://devicetree.org/schemas/soc/zte/zte,zx279133-pon-route.yaml#
+$schema: http://devicetree.org/meta-schemas/core.yaml#
+
+title: ZTE ZX279133 PON SerDes route control
+
+maintainers:
+  - cnjn <1779895255@qq.com>
+
+description:
+  The route control selects whether the PON SerDes is connected to XMAC1.
+  It is separate from the PON protocol register block and is consumed by the
+  NPPT Ethernet controller for the direct WAN path.
+
+properties:
+  compatible:
+    items:
+      - const: zte,zx279133-pon-route
+      - const: syscon
+
+  reg:
+    maxItems: 1
+
+required:
+  - compatible
+  - reg
+
+additionalProperties: false
+
+examples:
+  - |
+    syscon@17000080 {
+        compatible = "zte,zx279133-pon-route", "syscon";
+        reg = <0x17000080 0x4>;
+    };
+
+...
diff --git a/MAINTAINERS b/MAINTAINERS
index 98d0f59668a680e2fc6042114049f78e7cf01b37..4d21da5fde86b70caa43a761e4e1c01efc321910 100644
--- a/MAINTAINERS
+++ b/MAINTAINERS
@@ -28412,6 +28412,7 @@ F:	Documentation/devicetree/bindings/clock/zte,zx279133-*.yaml
 F:	Documentation/devicetree/bindings/gpio/zte,zx279133-gpio.yaml
 F:	Documentation/devicetree/bindings/hwmon/zte,zx279133-pvt.yaml
 F:	Documentation/devicetree/bindings/mfd/zte,zx279133-wdt-reset.yaml
+F:	Documentation/devicetree/bindings/net/dsa/zte,zx279133-*.yaml
 F:	Documentation/devicetree/bindings/net/zte,sr1010-rtl8372n-id-audit.yaml
 F:	Documentation/devicetree/bindings/net/zte,sr1010-zx279051-audit.yaml
 F:	Documentation/devicetree/bindings/net/zte,zx279133-mdio.yaml
@@ -28420,6 +28421,7 @@ F:	Documentation/devicetree/bindings/nvmem/zte,zx279133-efuse.yaml
 F:	Documentation/devicetree/bindings/phy/zte,zx279133-*.yaml
 F:	Documentation/devicetree/bindings/pinctrl/zte,zx279133-pinctrl.yaml
 F:	Documentation/devicetree/bindings/reset/zte,zx296718-reset.yaml
+F:	Documentation/devicetree/bindings/soc/zte/zte,zx279133-*.yaml
 F:	Documentation/devicetree/bindings/soc/zte,zx279133-pwm-audit.yaml
 F:	Documentation/devicetree/bindings/soc/zte,zx279133-xpcs-nppt-audit.yaml
 F:	drivers/soc/zte/
